pkg/network/ovn_kubernetes.go

```diff
@@ -473,6 +458,36 @@ type replicaCountDecoder struct {
 	} `json:"controlPlane"`
 }
 
+// bootstrapOVNGatewayConfig sets the Network.operator.openshift.io.Spec.DefaultNetwork.OVNKubernetesConfig.GatewayConfig value
+// based on the values from the "gateway-mode-config" map if any
+func bootstrapOVNGatewayConfig(conf *operv1.Network, kubeClient client.Client) {
+	// handle upgrade logic for gateway mode in OVN-K plugin (migration from hidden config map to using proper API)
+	// TODO: Remove this logic in future releases when we are sure everyone has migrated away from the config-map
+	cm := &corev1.ConfigMap{}
+	nsn := types.NamespacedName{Namespace: "openshift-network-operator", Name: "gateway-mode-config"}
+	err := kubeClient.Get(context.TODO(), nsn, cm)
+	modeOverride := OVN_SHARED_GW_MODE
+	routeViaHost := false
+
+	if err != nil {
+		klog.Infof("Did not find gateway-mode-config. Using default gateway mode: %s", OVN_SHARED_GW_MODE)
+	} else {
+		modeOverride = cm.Data["mode"]
+		if modeOverride != OVN_SHARED_GW_MODE && modeOverride != OVN_LOCAL_GW_MODE {
+			klog.Warningf("gateway-mode-config does not match %q or %q, is: %q. Using default gateway mode: %s",
+				OVN_LOCAL_GW_MODE, OVN_SHARED_GW_MODE, modeOverride, OVN_SHARED_GW_MODE)
+			modeOverride = OVN_SHARED_GW_MODE
+		}
+	}
+	if modeOverride == OVN_LOCAL_GW_MODE {
+		routeViaHost = true
+	}
+	conf.Spec.DefaultNetwork.OVNKubernetesConfig.GatewayConfig = &operv1.GatewayConfig{
```

Comment:
This doesn't actually update the config map right? Do we need to update the config map from here so that when we remove the ability to use the hidden configmap we dont break users? Or should we print some warning or event to tell the user they need to update the gatewayConfig in their configmap?

Reply:
We do update the configmap when this function returns. Let me explain:

1. https://github.com/openshift/cluster-network-operator/blob/38b07c3f025b0ab7dad6b5480605b20976fe1e47/pkg/network/bootstrap.go#L18
2. https://github.com/openshift/cluster-network-operator/blob/38b07c3f025b0ab7dad6b5480605b20976fe1e47/pkg/network/ovn_kubernetes.go#L485
3. https://github.com/openshift/cluster-network-operator/blob/38b07c3f025b0ab7dad6b5480605b20976fe1e47/pkg/network/ovn_kubernetes.go#L284
4. Within `bootstrapOVNConfig` we do our thing. 
5. Then we return all the way back to the bootstrap call https://github.com/openshift/cluster-network-operator/blob/38b07c3f025b0ab7dad6b5480605b20976fe1e47/pkg/controller/operconfig/operconfig_controller.go#L192
6.  Aaaand this is where we update the CRD: https://github.com/openshift/cluster-network-operator/blob/38b07c3f025b0ab7dad6b5480605b20976fe1e47/pkg/controller/operconfig/operconfig_controller.go#L200-L201

Hence any changes done to the CRD during bootstrap gets written into the CRD. After all that the rendering starts and it regards the updated CRD the source of truth.


I tested it using the steps mentioned in https://docs.google.com/document/d/1SYaNQNW4sWVTE9XgHG6gKsS4ZActK8_Z7wILXU5YuoU/edit#heading=h.d5s838mpuww0 in a cluster-bot cluster and it works well. The CRD does get updated to something like:
```
Gateway Config:{
    Routing Via Host: true
}
```